libeos-updater-util/avahi-service-file.c

```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+                         "q", &index);
+
+  return index;
+}
+
+static guint16
+get_avahi_service_port (GVariantDict *options_dict)
+{
+  // TODO: Should we store the port number in the configuration
+  // instead having it as the compile-time constant? I can't remember
+  // why I opted for this solution (was it lack of the configuration
+  // file?)
+  //
+  // In case when config file doesn't specify the port number and the
+  // number wasn't provided with the options_dict variant, likely bail
+  // out.
+  guint16 port = EOS_AVAHI_PORT;
+
+  g_variant_dict_lookup (options_dict, EOS_OSTREE_AVAHI_OPTION_PORT_Q,
+                         "q", &port);
+
+  return port;
+}
+
+static gboolean
+fill_filter_with_refs (OstreeBloom   *filter,
+                       OstreeRepo    *repo,
+                       GCancellable  *cancellable,
+                       GError       **error)
+{
+  g_autoptr(GHashTable) refs = NULL;
+  GHashTableIter iter;
+  gpointer key;
+
+  if (!ostree_repo_list_refs (repo,
+                              NULL, /* no prefix filtering */
+                              &refs,
+                              cancellable,
+                              error))
+    return FALSE;
+
+  g_hash_table_iter_init (&iter, refs);
+  while (g_hash_table_iter_next (&iter, &key, NULL))
+    {
+      ostree_bloom_add_element (filter, key);
+    }
+
+  return TRUE;
+}
+
+#define variant_sink_new(format_string, ...) (g_variant_ref_sink (g_variant_new (format_string, __VA_ARGS__)))
+
+static GVariant *
+get_bloom_filter_variant (OstreeRepo *repo,
+                          GVariantDict *options_dict,
+                          GCancellable *cancellable,
+                          GError **error)
+{
+  guint32 bloom_size = 255 - strlen(EOS_OSTREE_AVAHI_V1_REFS_BLOOM_FILTER_FIELD) - 2;
+  guint8 bloom_k = 1;
+  guint8 bloom_hash_id = EOS_OSTREE_AVAHI_STR_BLOOM_HASH_ID;
+  OstreeBloomHashFunc bloom_hash_func;
+  g_autoptr(OstreeBloom) filter = NULL;
+  g_autoptr(GBytes) sealed_filter = NULL;
+  g_autoptr(GVariant) bloom_bits_variant = NULL;
+
+  g_variant_dict_lookup (options_dict, EOS_OSTREE_AVAHI_OPTION_BLOOM_SIZE_U,
+                         "u", &bloom_size);
+  g_variant_dict_lookup (options_dict, EOS_OSTREE_AVAHI_OPTION_BLOOM_K_Y,
+                         "y", &bloom_k);
+  g_variant_dict_lookup (options_dict, EOS_OSTREE_AVAHI_OPTION_BLOOM_HASH_ID_Y,
+                         "y", &bloom_hash_id);
+
+  switch (bloom_hash_id)
+    {
+    case EOS_OSTREE_AVAHI_STR_BLOOM_HASH_ID:
+      bloom_hash_func = ostree_str_bloom_hash;
+      break;
+
+    default:
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                   "unknown bloom hash function id %u", bloom_hash_id);
+      return NULL;
+    }
+
+  filter = ostree_bloom_new (bloom_size, bloom_k, bloom_hash_func);
+  if (!fill_filter_with_refs (filter, repo, cancellable, error))
+    return FALSE;
+  sealed_filter = ostree_bloom_seal (filter);
```

Comment:
Maybe consider also adding an API for getting the `k` parameter and a hash function from `OstreeBloom`. I could use it here.